[redis] Publish Filter-Id value on redis

butjar opened the pull request:
# Description
This PR allows to pass "QoS profiles" to accel-ppp via `Filter-Id` attributes in radius *Access-Accept* messages. The value of `Filter-Id` is stored in accel-ppp's internal session struct and sent as `qos_val` attribute of the JOSN messages published on the redis bus.
# Test instructions
## Publishing `Filter-Id` values to redis

### Requirements
- Radius setup with `Filter-Id` Attributes returned on *Access-Accept* e.g. by adding `Filter-Id` entries to the `radreply` table for freeradius with `rlm_sql` (https://wiki.freeradius.org/guide/SQL-HOWTO-for-freeradius-3.x-on-Debian-Ubuntu#populating-sql)
- An accel-ppp instance connected to a running redis-server
- `redis-cli`
- `accel-ppp.conf` with `redis` module enabled and a config similar to the following (with **ev_radius_access_accept=yes** enabled):
    ```
    ...
    redis
    ...
    [redis]
    host=redis
    port=6379
    pubchan=accel-ppp

    # select the event types to emit a message via redis
    ev_ses_pre_finished=yes
    ev_ses_acct_start=yes
    ev_radius_access_accept=yes
    ```
  
### Test steps
1. Connect `redis-cli` to the running redis instance and subscribe to the `accel-ppp` channel:
    ```
    $ redis-cli 
        127.0.0.1:6379> SUBSCRIBE accel-ppp
        Reading messages... (press Ctrl-C to quit)
        1) "subscribe"
        2) "accel-ppp"
        3) (integer) 1

    ```
2. Create a client session with a user that has a `Filter-Id`:
    ```
    $ pppd pty "pppoe -I <IFACE> -T 80 -U -m 1350" \
        noccp \
        ipparam <IFACE> \
        linkname <IFACE> \
        noipdefault \
        noauth \
        default-asyncmap \
        defaultroute \
        hide-password \
        updetach \
        mtu 1350 \
        mru 1350 \
        noaccomp \
        nodeflate \
        nopcomp \
        novj \
        novjccomp \
        lcp-echo-interval 40 \
        lcp-echo-failure 3 \
        user <PPP_USER> \
        password <PPP_PASWORD>
    ```

### Expected behavior
The `Filter-Id` should be published in the messages on the `accel-ppp` channel as `qos_val`:

```
127.0.0.1:6379> SUBSCRIBE accel-ppp
Reading messages... (press Ctrl-C to quit)
1) "subscribe"
2) "accel-ppp"
3) (integer) 1
1) "message"
2) "accel-ppp"
3) "{ \"event\": \"radius-access-accept\", \"ctrl_type\": \"pppoe\", \"channel_name\": \"aa:aa:aa:aa:00:c8\", \"session_id\": \"d89da8604fc10e66\", \"called_station_id\": \"06:9d:98:10:37:7e\", \"calling_station_id\": \"aa:aa:aa:aa:00:c8\", \"name\": \"pppoe\", \"ip_addr\": \"\", \"pppoe_sessionid\": 1, \"ctrl_ifname\": \"vethAC.7.200\", \"nas_identifier\": \"nas1\", \"qos_val\": \"1\\/16000\\/2000\" }"
1) "message"
2) "accel-ppp"
3) "{ \"event\": \"session-acct-start\", \"ctrl_type\": \"pppoe\", \"channel_name\": \"aa:aa:aa:aa:00:c8\", \"session_id\": \"d89da8604fc10e66\", \"called_station_id\": \"06:9d:98:10:37:7e\", \"calling_station_id\": \"aa:aa:aa:aa:00:c8\", \"name\": \"pppoe\", \"username\": \"sp-tin\", \"ip_addr\": \"192.51.0.3\", \"pppoe_sessionid\": 1, \"ctrl_ifname\": \"vethAC.7.200\", \"nas_identifier\": \"nas1\", \"qos_val\": \"1\\/16000\\/2000\" }"
```

butjar committed:
[docu] State purpose of new ap_session members

Add some inline comments to clarify why new members have been inserted
into the `ap_session` struct. New members:
- `conn_pppoe_sid`
- `filter_id`

rr-santos commented:
We do not need to add new functionalities to the redis module, other than just publishing messages with the fields present in the `ap_session` object.

Similarly to the pppoe session id, where we populate the session object inside the pppoe module, the filter id should be processed when handling the access-accept attributes from radius: https://github.com/dpdk-vbng-cp/accel-ppp/blob/master/accel-pppd/radius/radius.c#L272 . There, we can even use the corresponding macro for the Filter-Id: https://github.com/dpdk-vbng-cp/accel-ppp/blob/master/accel-pppd/radius/attr_defs.h#L14 .

This also removes the need to subscribe to the access accept events in the redis module in order to handle the filter-id attribute, as currently we also are not interested in receiving these messages on the corresponding redis subscriber.

butjar commented:
> We do not need to add new functionalities to the redis module, other than just publishing messages with the fields present in the `ap_session` object.
> 
> Similarly to the pppoe session id, where we populate the session object inside the pppoe module, the filter id should be processed when handling the access-accept attributes from radius: https://github.com/dpdk-vbng-cp/accel-ppp/blob/master/accel-pppd/radius/radius.c#L272 . There, we can even use the corresponding macro for the Filter-Id: https://github.com/dpdk-vbng-cp/accel-ppp/blob/master/accel-pppd/radius/attr_defs.h#L14 .
> 
> This also removes the need to subscribe to the access accept events in the redis module in order to handle the filter-id attribute, as currently we also are not interested in receiving these messages on the corresponding redis subscriber.

You suggest to implement the "new functionality" in accel-ppp itself then. I see that it helps to reduce the events we need to subscribe to.  However changing code outside our module will increase the complexity of keeping our fork up to date with upstream. Decoupling between the events we publish and the events we need to process should solve this problem as well.

Regarding the pppoe session id, I would also prefer not to add it as a `ap_session` member since a lookup should be possible when we need it (see https://github.com/xebd/accel-ppp/blob/e9d369aa0054312b7633e964e9f7eb323f1f3d69/accel-pppd/ctrl/pppoe/pppoe.c#L224-L225):
```
	struct ppp_t *ppp = container_of(ses, typeof(*ppp), ses);
	struct pppoe_conn_t *conn = container_of(ppp, typeof(*conn), ppp);
```

Similarly, I think the radius replies might be available for a lookup in the `pd_list` member of `ap_session` (https://github.com/dpdk-vbng-cp/accel-ppp/blob/master/accel-pppd/include/ap_session.h#L106).

rr-santos commented:
The main reason behind adding the session attribute in the radius module was to keep the application logic consistent. If the logic of processing the radius attributes when handling the radius accept messages is already there, it's intuitive that we add the new radius attribute in the same place (contrary to have this scattered around different places in the code).

Separating our fork's code changes between the upstream code is a different issue that needs to be handled separately (e.g. the pppoe session ID modifications in the pppoe module would also need to be changed).

butjar commented:
> The main reason behind adding the session attribute in the radius module was to keep the application logic consistent. If the logic of processing the radius attributes when handling the radius accept messages is already there, it's intuitive that we add the new radius attribute in the same place (contrary to have this scattered around different places in the code).

I agree that it would be consistent to what we have. But we scatter the implementation more if we touch the radius functionality.

butjar committed:
radius: Insert Filter-Id into session context

Insert `Filter-Id` into `ap_session` struct. The `rad_proc_attrs`
function is called on *Access-Accept* and parses attributes from the
radius message. Some of the parsed attributes are already persisted in
`ap_session`. This commit writes the value of the `Filter-Id` attribute
into a newly introduced member of the `ap_session`. Persisting the value
in the session context provides access to the `Filter-Id` value for
later processing (e.g. most trition events pass a `ap_session` pointer
to the event listeners).

butjar committed:
redis: Map Filter-Id into qos_val in redis message

This commit adds the `qos_val` attribute into published redis messages.
For now we map the `Filter-Id` attribute received from a radius
*Access-Accept* into the `qos_val`.

rr-santos commented:
Just a minor thing, I'd also add freeing the allocated memory or the filter_id attribute in https://github.com/dpdk-vbng-cp/accel-ppp/blob/master/accel-pppd/session.c#L172 . Apart from that, looks good!

butjar commented:
> Just a minor thing, I'd also add freeing the allocated memory or the filter_id attribute in https://github.com/dpdk-vbng-cp/accel-ppp/blob/master/accel-pppd/session.c#L172 . Apart from that, looks good!

Good catch

butjar commented on accel-pppd/session.c:
```diff
@@ -246,6 +246,10 @@ void __export ap_session_finished(struct ap_session *ses)
+	if (ses->filter_id)
+		_free(ses->filter_id);
+		ses->filter_id = NULL;
+
```
I tested this code locally with GDB by terminating the session:
```
[2021-02-26 10:42:16]: debug: ppp0: pppoe: ppp finished
[2021-02-26 10:42:16]: debug: ap_redis: Epoll fd content: 1.
[Switching to Thread 0x7ffff296c700 (LWP 6681)]

Thread 7 "accel-pppd" hit Breakpoint 1, ap_session_finished (ses=0x7fffdc001c70)
    at /home/butjar/bisdn/vbng/accel-ppp/accel-pppd/session.c:249
249             if (ses->filter_id)
(gdb) p ses->filter_id
$1 = 0x7fffe80072f0 "1/16000/2000"  <-------------------  Before freeing the pointer
(gdb) l
244                     ses->net->release(ses->net);
245
246             if (ses->timer.tpd)
247                     triton_timer_del(&ses->timer);
248
249             if (ses->filter_id)
250                     _free(ses->filter_id);
251                     ses->filter_id = NULL;
252
253     #ifdef USE_BACKUP
(gdb) n
250                     _free(ses->filter_id);
(gdb) n
[2021-02-26 10:42:43.014] ap_redis: epoll_ctl failed: 4 (Interrupted system call)
[2021-02-26 10:42:43]: error: ap_redis: epoll_ctl failed: 4 (Interrupted system call)
251                     ses->filter_id = NULL;
(gdb) n
[2021-02-26 10:42:43.895] ap_redis: epoll_ctl failed: 4 (Interrupted system call)
[2021-02-26 10:42:43]: error: ap_redis: epoll_ctl failed: 4 (Interrupted system call)
258             if (ap_shutdown && !ap_session_stat.starting && !ap_session_stat.active && !ap_session_stat.finishing) {
(gdb) p ses->filter_id
$2 = 0x0  <-------------------  After freeing the pointer
```

micron10 commented:
Hi  @butjar 

i try but in our test case use : attr=Mikrotik-Rate-Limit

How to set, i see in this case you use Filter-Id

In original accel code in shaper.c have this : 

static void parse_attr(struct rad_attr_t *attr, int dir, int *speed, int *burst, int *tr_id)
{
        if (attr->attr->type == ATTR_TYPE_STRING) {
                int vendor = attr->vendor ? attr->vendor->id : 0;
                if (vendor == 9) {
                        /* VENDOR_Cisco */
                        parse_string_cisco(attr->val.string, dir, speed, burst, tr_id);
                } else if (vendor == 14988 && attr->attr->id == 8) {
                        /* VENDOR_Mikrotik && Mikrotik-Rate-Limit */
                        dir = (dir == ATTR_DOWN) ? ATTR_UP : ATTR_DOWN;
                        parse_string_simple(attr->val.string, dir, speed, burst, tr_id);
                } else
                        parse_string(attr->val.string, dir, speed, burst, tr_id);
        } else if (attr->attr->type == ATTR_TYPE_INTEGER)
                *speed = conf_multiplier * attr->val.integer;
}

butjar commented:
> Hi @butjar
> 
> i try but in our test case use : attr=Mikrotik-Rate-Limit
> 
> How to set, i see in this case you use Filter-Id
> 
> In original accel code in shaper.c have this :
> 
> static void parse_attr(struct rad_attr_t *attr, int dir, int *speed, int *burst, int _tr_id) { if (attr->attr->type == ATTR_TYPE_STRING) { int vendor = attr->vendor ? attr->vendor->id : 0; if (vendor == 9) { /_ VENDOR_Cisco _/ parse_string_cisco(attr->val.string, dir, speed, burst, tr_id); } else if (vendor == 14988 && attr->attr->id == 8) { /_ VENDOR_Mikrotik && Mikrotik-Rate-Limit */
> dir = (dir == ATTR_DOWN) ? ATTR_UP : ATTR_DOWN;
> parse_string_simple(attr->val.string, dir, speed, burst, tr_id);
> } else
> parse_string(attr->val.string, dir, speed, burst, tr_id);
> } else if (attr->attr->type == ATTR_TYPE_INTEGER)
> *speed = conf_multiplier * attr->val.integer;
> }

We support only `Filter-Id` so far. I think this PR demonstrates how this can be done quite clearly. `Mikrotic-Rate-Limit` could be similarly parsed as `Filter-Id`, see: [radius.c#L387](https://github.com/dpdk-vbng-cp/accel-ppp/blob/8dd9bb73815f86b3ce418ea02a86ae3110e70c9b/accel-pppd/radius/radius.c#L387). You can reuse the redis implementation for the `Filter-Id` by writing the content of the `Mikrotic-Rate-Limit` attribute to the [`filter_id` member](https://github.com/dpdk-vbng-cp/accel-ppp/blob/afb19585d2ff4c61bef078f0bc92a2359f33d76d/accel-pppd/include/ap_session.h#L125) of the `ap_session` struct (in `radius.c`: `rpd->ses->filter_id`). 
Of course you could hook into the shaper as well.

micron10 commented:
Hi yes i see use only Filter-Id , i try to make change Filter-Id to Mikrotik-Rate-Limit it did not work.

Problem is Mikrotik is external dictionary.

Fitler_Id is come with attr_defs.h (Filter_Id is default attr )

and in this case is not options to import external dictionary in code or change Filter_Id to Mikrotik-Rate-Limit.

in PR #10 your parser is great only on part where is final return is make a problem : 

+/* This function parses the QoS value from radius Access-Accept messages. At
+ *  * the moment we parse the attribute Filter-Id similar as in
+ *   * shaper.c:check_radius_attrs. A value of the format
+ *    * <SERVICE_TYPE>/<DL_RATE_KBITS>/<UL_RATE_KBITS> is expected. */
+static const char* parse_qos_val(struct rad_packet_t *pack)
+{
+       struct rad_attr_t *attr;
+
+       list_for_each_entry(attr, &pack->attrs, entry) {
+               if (attr->vendor && attr->vendor->id != conf_vendor)
+                       continue;
+               if (!attr->vendor && conf_vendor)
+                       continue;
+               if (attr->attr->id != conf_attr_down && attr->attr->id != conf_attr_up)
+                       continue;
+
+       return attr->val.string;
+       }
+
+    return NULL;
+}


This : return NULL is need to fix .

i try and after remove NULL receive this in redis : 

3) "{ \"event\": \"session-acct-start\", \"ctrl_type\": \"pppoe\", \"channel_name\": \"1e:45:a7:fe:40:06\", \"session_id\": \"9b783cb1f6e7b0ee\", \"called_station_id\": \"8e:f7:3e:82:72:17\", \"calling_station_id\": \"1e:45:a7:fe:40:06\", \"name\": \"pppoe\", \"username\": \"inactiveuser\", \"ip_addr\": \"172.16.0.3\", \"pppoe_sessionid\": 64, \"ppp_name\": \"ppp0\", \"ctrl_ifname\": \"eth1\", \"nas_identifier\": \"test-01\", \"qos_val\": \"H\x92\" }"


when is login with Normal user : 

3) "{ \"event\": \"session-acct-start\", \"ctrl_type\": \"pppoe\", \"channel_name\": \"1e:45:a7:fe:40:06\", \"session_id\": \"9b783cb1f6e7b0ef\", \"called_station_id\": \"8e:f7:3e:82:72:17\", \"calling_station_id\": \"1e:45:a7:fe:40:06\", \"name\": \"pppoe\", \"username\": \"normaluser\", \"ip_addr\": \"10.26.0.1\", \"pppoe_sessionid\": 128, \"ppp_name\": \"ppp0\", \"ctrl_ifname\": \"eth1\", \"nas_identifier\": \"test-01\", \"qos_val\": \"102400k\\/102400k 112640k\\/112640k\" }
